### PR TITLE
Terminal Widget based on xtermjs

### PR DESCRIPTION
**Pull request by MarcSkovMadsen**
https://github.com/holoviz/panel/issues/1925

Implements a Terminal Component based on [xtermjs](https://github.com/xtermjs/xterm.js/)

## Must Do

[x] Make it work in Jupyter Hub (See image with issue in post below).
[x] Create Reference Guide Notebook
[x] Design the api
[x] Support input
[x] Support output including links, emojis and danish characters.
[x] Fix size issues (using add on [fit](https://github.com/xtermjs/xterm.js/tree/master/addons/xterm-addon-fit))
[x] Define `.js` and `.css` assets to include on bokeh model.
[x] Make it Work on Server
[x] Make it Work in notebook

# Nice to Have

[] Support running subprocesses like bash, python and ipython in pty
    - [] On windows (winpty?)
    - [] Improve on `add_periodic_callback`. Should this be seperate thread or other period length?
    - [x] On Linux
    - [x] Handle user exit from terminal.
[] Support Addons
    - [] [Attach](https://github.com/xtermjs/xterm.js/tree/master/addons/xterm-addon-attach)
    - [x] [fit](https://github.com/xtermjs/xterm.js/tree/master/addons/xterm-addon-fit)
    - [] [Search](https://github.com/xtermjs/xterm.js/tree/master/addons/xterm-addon-search)
    - [x] [Web-Links](https://github.com/xtermjs/xterm.js/tree/master/addons/xterm-addon-web-links)
[] Make `.fit` calculation in `.ts` model 100% bullet proof and also support `sizing_mode="stretch_height"`.
[] Determine whether to inherit from `StreamIO` or not. c.f. Philipp comment below.
[] Fix hacks for handling repeated input/ output
[] Enable copy-paste
[] Implement nice kill in notebook (see image in posts below).
[x] Example of using subprocess.Popen
[x] Example using streaming data from some longer running process
[x] Example of redirecting python log

### CODE REVIEW

**Comment by MarcSkovMadsen** [comment]
I just added the basic "Button Extension" described in the [Custom Extension guide](https://github.com/holoviz/panel/pull/2084). It made is SO EASY to get started with a setup. It took a few minutes to get a basic custom extension working. 

Maybe a `panel create-extension` cli command doing the same could make it even more simple?

**Comment by MarcSkovMadsen** [comment]
A minimum POC is there

https://user-images.githubusercontent.com/42288570/111262373-78e2b480-8624-11eb-8e10-92dc3fa8c7d1.mp4

**Comment by MarcSkovMadsen** [comment]
@philippjfr . If you have good ideas about the api we want feel free to provide an example.

I will have to explore and learn because I have not working with terminals before. And not so much with redirecting logs, stdout etc. or using pty.

**Comment by philippjfr** [comment]
It should probably just implement the file protocol, i.e. at minimum implement `write` but maybe the entire interface including `read`, `seek`, `readline`, `readlines` `readable`, `writable`, `seekable`, `tell`, `close`, `closeable`, `flushed` etc. Probably just start with `write` and then do:

```python
import sys

term = Terminal()
sys.stdout = term
sys.stderr = term
```

**Comment by MarcSkovMadsen** [comment]
Got the basics working. The Terminal is now defined as

```python
class Terminal(StringIO, Widget):
    # Parameters to be mapped to Bokeh model properties
    value = param.String(doc="""
        User input from the terminal""")
    object = param.String(doc="""
        System output to the terminal""")
```

- I was in doubt. But I believe the `Terminal` is a widget with a `value` (~user input) and an `object` (output) parameter.
- I will have to learn whether to inherit from StringIO is a good thing or not. For now it works.
- For now I have defined the value and object as `String`. I don't know if it should have been `Bytes`. I have a feeling that it is needed for some uses cases. For example `term.write('\x1bc')` does not clear the screen as described here https://github.com/xtermjs/xterm.js/issues/950

**Comment by philippjfr** [comment]
>I was in doubt. But I believe the Terminal is a widget with a value (~user input) and an object (output) parameter.

Definitely straddling the line but if it's accepting user input, it's a widget.

>I will have to learn whether to inherit from StringIO is a good thing or not. For now it works.

Don't think this is a good idea. Just satisfy the file-protocol.

>For now I have defined the value and object as String parameters. I don't know if it should have been Bytes. I have a feeling that it is needed for some uses cases. For example term.write('\x1bc') does not clear the screen as described here xtermjs/xterm.js#950

Not sure about separating the `value`/`object` naming. Also what does `object` return? The entire history? The last output?

File: panel/models/terminal.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// See https://docs.bokeh.org/en/latest/docs/reference/core/properties.html
+import * as p from "@bokehjs/core/properties"
+import { div } from "@bokehjs/core/dom";
+
+
+// The view of the Bokeh extension/ HTML element
+// Here you can define how to render the model as well as react to model changes or View events.
+export class TerminalView extends HTMLBoxView {
+    model: Terminal
+    term: any // Element
```
Comment (Tyriar) [inline panel/models/terminal.ts:13]: Not sure about your build/module system but you might be able to do this to bring the type in without the implementation:

```ts
import type { Terminal } from 'xterm';
```

File: panel/tests/widgets/test_terminal.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import panel as pn
+import sys
+import subprocess
+import shlex
+
+pn.config.js_files["xtermjs"]="https://unpkg.com/xterm@4.11.0/lib/xterm.js"
+pn.config.css_files.append("https://unpkg.com/xterm@4.11.0/css/xterm.css")
```
Comment (philippjfr) [inline panel/tests/widgets/test_terminal.py:7]: Assume this is for testing? These should be declared on the model.

Comment (MarcSkovMadsen) [reply panel/tests/widgets/test_terminal.py:7]: Sure. One step at a time :-)

File: panel/models/terminal.ts (hunk 1 above)
Comment (MarcSkovMadsen) [reply panel/models/terminal.ts:13]: Thanks. I will try it out.

**Comment by MarcSkovMadsen** [comment]
The object returns the last output.

The value returns the last input from the user

**Comment by MarcSkovMadsen** [comment]
Status: Output works, layout works (with a few hacks)
Next Step: User Input

Uploading panel_terminal3.mp4…

**Comment by MarcSkovMadsen** [comment]
Got the `.js` and `.css` imports working in the notebook. And started writing the reference guide.

image

**Comment by MarcSkovMadsen** [comment]
When "streaming to the terminal" in the notebook I experienced

**Comment by MarcSkovMadsen** [comment]
When I kill the process in the notebook it looks like

image

**Comment by MarcSkovMadsen** [comment]
Importing the `terminal` extension in Jupyterlab does not work

image

**Comment by philippjfr** [comment]
I would seriously discourage displaying anything in the same cell where you are loading the extension.

**Comment by philippjfr** [comment]
That seems to be a bug though, it should wait on bokeh to be loaded instead of erroring. But give it a few seconds before you display something to let Bokeh load. If that works then it's a distinct bug and you shouldn't worry about it.

**Comment by MarcSkovMadsen** [comment]
I just used one cell to make it as "minimal" as possible. It's the same problem when using multiple cells @philippjfr 

image

**Comment by MarcSkovMadsen** [comment]
When I don't import the `terminal` extension it works. 

image

The problem for me is that importing the terminal works in the notebook.

**Comment by philippjfr** [comment]
I'll take a look when I get a chance.

**Comment by MarcSkovMadsen** [comment]
Thanks @philippjfr . It's not urgent as I will just finalize the rest and then you can take a look as a part of the review/ rewrite you will be doing anyways. Thanks.

**Pull request merged**